qiskit/qasm3/ast.py

```diff
@@ -679,17 +679,19 @@ class ForLoopStatement(Statement):
             | "[" Range "]"
     """
 
-    __slots__ = ("indexset", "parameter", "body")
+    __slots__ = ("indexset", "parameter", "body", "type")
 
     def __init__(
         self,
         indexset: Union[Identifier, IndexSet, Range],
         parameter: Identifier,
         body: ProgramBlock,
+        type: ClassicalType,
```

Comment:
while `qasm3.ast.ForLoopStatement` is not part of the public API, I think it would be better to kvarg `type=None` and raise a deprecation warning when not set.